### ISSUE TITLE
Git on Windows client corrupts files > 4Gb

### ISSUE THREAD

**Issue by obe1line** [git-lfs/git-lfs#2434]
When cloning a file larger than 4Gbyte from a BitBucket server repository (LFS enabled), the file is not reconstructed correctly E.g. a 6Gb file results in a 700Mb file. The lfs/objects folder contains the correct file however.

Steps to reproduce:

Server:  Basic install of Debian with Bitbucket Server 4.6, Git 2.13 (64-bit)
Client 1:  Ubuntu 16.02 64 bit, Git 2.13 and Git-LFS 2.2.1 (both 64-bit)
Client 2:  Windows 64 bit (2012), Git 2.13 and Git-LFS 2.2.1 (both 64-bit)

1. Create a repository on the server with LFS support enabled
2. Clone the repository (git clone ssh://git@server:7999/tst/test-git.git)
3. git track '*.iso' , commit and push to the remote
4. add an iso file larger than 4Gb (I used Visual Studio 2013 Update 4 which is 5.82Gb), commit and push to the remote
5. Clone the repository into a clean folder

Client 1 works correctly
Client 2 pulls down the file correctly into .git/lfs/objects/aa/6d/aa6d2a8e9acbb78895b3d2c6ae3cb0db737344aa82b2859d31f757deec931049 but does not reconstruct/copy it correctly to the destination folder (it results in a 1.82Gb file).

Atlassian have looked into the problem and believe that the BitBucket server is working correctly, due to the fact that the correct content is retrieved over the network into the temporary object file (the CRCs match the original file).

Note that no Git configuration has changed (smudge filters etc are the default).

If the file is removed and "git lfs pull" performed, the file is created correctly. Using "git lfs clone" also works.

**Comment by technoweenie** [comment]
First, you can confirm that BitBucket is sending the data correctly by checking the file in `.git/lfs/objects`. If this checks out, the server is doing things correctly:

```sh
$ cd .git/lfs/objects/aa/6d
$ shasum -a 256 aa6d2a8e9acbb78895b3d2c6ae3cb0db737344aa82b2859d31f757deec931049
aa6d2a8e9acbb78895b3d2c6ae3cb0db737344aa82b2859d31f757deec931049  aa6d2a8e9acbb78895b3d2c6ae3cb0db737344aa82b2859d31f757deec931049
```

Next, the fact that it works with `git lfs clone` and `git lfs pull` means the Git LFS code for copying files works. This ultimately comes down to [`lfs.LinkOrCopy()`](https://github.com/git-lfs/git-lfs/blob/ce66ce324774187949ee562afe25923e23734acc/lfs/util.go#L189-L223), which either makes a hard link (if `.git/lfs/objects` is on the same partition as your working dir) or manually writes the bytes to the new location.

So, this leaves the git filters. There are two modes that could be causing problems:

1. The `process` mode is default in your reported git and lfs versions. The `process` mode uses a protocol to receive smudge requests from Git via STDIN, and to deliver object contents from LFS via STDOUT.
2. The older `smudge` mode is basically a request from Git to get the contents for a single file. It may be worth triggering the smudge filter directly to see if it does the same thing:

```sh
$ GIT_LFS_SKIP_SMUDGE=1 git clone https://git-server/your/repo
$ cd repo
$ cat path/to/file
version https://git-lfs.github.com/spec/v1
oid sha256:98ea6e4f216f2fb4b69fff9b3a44842c38686ca685f3f55dc48c5d3fb1107be4
size 3

$ cat path/to/file | git lfs smudge > smudged-file.bin
$ shasum -a 256 hi.txt
98ea6e4f216f2fb4b69fff9b3a44842c38686ca685f3f55dc48c5d3fb1107be4  hi.txt
```

Based on that, I think one of the following could be happening:

1. Git LFS is not sending the correct contents through the filter process. I imagine this would've caught by now, as it would affect every object.
2. The file contents aren't being piped through `STDOUT` correctly. Maybe there are some special characters in your files that are stripped or something?
3. Git is not receiving the file connects correctly. You can confirm this by running `GIT_TRACE_PACKET=1 git clone` or something, but the output would be _massive_ for your files :)

Some questions:

* Are you experiencing this with a specific type of file, or just files over a certain size?
* Are the corrupt files being truncated? If a 6GB file outputs a 700MB file to your working directory, does it match the first 700MB of the 6GB file?

It'd be really helpful if we could get a sample file that exhibits this behavior. I imagine that's a no-go, so we may have to come up with a special build of LFS with special tracing powers. @ttaylorr, any thoughts? Did I miss any debugging questions or trial commands to run?

**Comment by ttaylorr** [comment]
@technoweenie that looks pretty comprehensive. My hunch is that it's related to one of the three issues you described as being process filter-related.

@obe1line do you have a copy of the file or repository that you could share? I think that would be the easiest way for me to debug this going forward.

**Comment by technoweenie** [comment]
Hey @obe1line, I ran this by a Git core dev, and he mentioned that Git on Windows does not support files over 4GB. Unfortunately, this is not something we can fix in LFS. The best we can do now add a warning when large objects are added.

As a workaround, I think you should disable smudging completely:

```sh
$ git lfs install --skip-smudge
$ git lfs env
... snip
git config filter.lfs.process = "git-lfs filter-process --skip"
git config filter.lfs.smudge = "git-lfs smudge --skip -- %f"
```

After that, you'll have to run `git lfs pull` any time you change branches or fetch updates from your remote.

@ttaylorr I think we should add an early warning in the filter smudge and process code, perhaps linking to a page offering this workaround.

**Comment by obe1line** [comment]
@technoweenie Thanks for the detailed comment.
In answer to your questions:

- All files over 4Gb are affected, not just specific types
- Yes, files are truncated with the 700Mb matching the original file
- Copying the file from .git/lfs/objects/ and renaming matches the original file

I used the GIT_TRACE_PACKET, GIT_TRACE and GIT_CURL_VERBOSE to produce output previously and yes, it is rather large (~12Gb from memory).

@ttaylorr It can be reproduced with any file >4Gb, nothing special about the repository. The file is pushed with the Linux client, and fetched with the Windows client - I assume if I had used Windows to push, then the file may not have transferred fully into the repository.

@technoweenie Is the workaround only applicable to fetch? i.e. would uploading a file via "git lfs push" work correctly even with the Git 4Gb limit?

**Comment by technoweenie** [comment]
> Is the workaround only applicable to fetch? i.e. would uploading a file via "git lfs push" work correctly even with the Git 4Gb limit?

It probably won't work on Windows. Added files are passed through the LFS `clean` filter (basically the reverse of the `smudge` filter) are probably subject to the same size limitation in Git.

**Comment by ndebard** [comment]
I'm seeing this on Ubuntu that is running on a Windows machine but is rebooted with Grub.  This may have something to do with the partition rather than windows itself?  See attached photo.
![20180301_021229](https://user-images.githubusercontent.com/4204146/36853030-8a9eafb4-1d2a-11e8-9060-cdb919e83e8d.jpg)

**Comment by ttaylorr** [comment]
Hi @ndebard -- thanks for commenting here. I think what your experiencing is the correct behavior, even though the output can be a bit confusing. This message appears when a file greater than 4.0 GB is copied into your working tree (it looks like `data/dump_006_ls_0.dat` is 9.0 GB). The reason this message occurs is to warn when checking out that same file on Windows, there may be issues with certain versions of Git.

Though the message isn't pertinent to you on Ubuntu (?), it will have relevancy for any colleagues of yours using Windows.

**Comment by dwall17** [comment]
Hi @ttaylorr 
   I'm very new to using Git so I just want to gain further clarification on your response here. If I'm correct, what you're saying is that this "malformed smudge" output isn't an error message, but instead a warning to Windows users pertaining to their version of Git? Considering that I'm using Ubuntu, would it be safe to simply ignore the message?

**Comment by ttaylorr** [comment]
> Considering that I'm using Ubuntu, would it be safe to simply ignore the message?

It is safe to ignore the message for yourself -- since you're not on Windows, your repository should work as expected even though it contains large files in the working copy. The warning is to remind you that a checkout of your repository _may not work_ on a Windows machine.

**Comment by luckydonald** [comment]
Could someone please clarify what this `smude` filter does, and how turning it off makes windows able to store bigger files?
Or is large file system not able to process large files over 4 GB at all? Would the common git be able to store them files?

**Comment by ttaylorr** [comment]
> Could someone please clarify what this `smude` filter does

Hi @luckydonald, thanks for asking! The `smudge` filter is applied to transform content from your index into your working copy. In practice, this means that we take the small reference (pointer) that LFS actually stores in your Git repository, and transparently convert it into a large file on your disk, so that it appears as if the large file itself is present in the repository.

> and how turning it off makes windows able to store bigger files?

I don't think that turning off the smudge filter would make Windows able to store bigger files. The issue is rather that Git has a limitation on Windows of not being able to correctly smudge files when the size of the outgoing content is larger than 4GB.

This isn't an inherent limit of the file system, rather an implementation detail of Git.

**Comment by ttaylorr** [comment]
> Basically, if you have 4GB files in your repo, you can clone it if you disable the smudge filter, but you can't commit or push?

Not quite. The issue is with 4 GiB files of _any source_, them coming from Git LFS is only one half of the problem. If another filter puts them there, or that's how they're stored in your Git repository, then it is not guaranteed that it will be checked out correctly by Git on Windows.

With regards to the Git LFS-part of that problem, if you have a >4 GiB LFS _object_ (read: not a Git object, but an LFS one), you can avoid introducing that into your local copy by passing `--exclude=path/to/file` (or `lfs.fetchexclude=path/to/file` in your .gitconfig). With either (or both) of these options passed, Git LFS will not download or check out the large file into your working copy, thus side-stepping the problem.

One thing that I think is important to remember, is that this issue does _not_ cause problems on Unix, macOS, or other platforms that don't have the >4 GiB file-size limitation. So, if you have a >4 GiB file in your repository (LFS or otherwise), it should work fine on platforms other than Windows. If you're on Windows, we are stuck with this behavior, so `--exclude` is the best way forward, IMHO.

**Comment by chowey** [comment]
For the record, I have been using Git LFS on Windows by disabling the smudge filter and the process filter. Files >4GB seem to work fine. It just means you need to manually `git lfs pull` any time you pull, switch branches, or clone a repo.

**Comment by JohnFrampton** [comment]
First: my colleges an me encounter that problem on windows 10 with newest NTFS filesystem which is without any doublt capable of handling files > 4 GB and even files of size up to 16Exabytes (see here http://www.ntfs.com/ntfs_vs_fat.htm). 

Second: the newest git (including lfs feature) is great.

Third: Thanx for all those proposals for avoiding the problem, but at the end we don't want to avoid to use/download/clone/pull files > 4 GB. 
We want git (lfs) to handle them as lfs files correctly. 

Is there a plan and timeframe to fix that problem on windows?

**Comment by bk2204** [comment]
So as I understand this issue, it's due to Git on Windows not supporting files greater than 4 GB properly.  The issue is that the smudge and clean filters are invoked by Git, and Git itself doesn't handle this gracefully.  Git LFS does handle this gracefully, but because it's invoked by Git (unless you disable the filters), the data is corrupted before it makes it to Git LFS.

To explain the issue with Git, it's because the Git codebase uses `unsigned long` for certain values.  On a 64-bit Unix system (including Linux and macOS), that type is 64 bits in length, and `unsigned long` is the canonical way to write a system-sized unsigned word type.  However, on Windows, `unsigned long` is always 32 bits.  Consequently, even a 64-bit Git on Windows doesn't handle large files.  There's an explanation of this issue in [a thread on the Git list](https://public-inbox.org/git/93ccf86d-0090-a311-9825-7e23fd6a4141@web.de/t/).

Git for Windows is already tracking this issue as git-for-windows/git#1063.  The good news is that when this is fixed in Git, everything should automatically work with any version of Git LFS.  In the mean time, there isn't anything we as Git LFS developers can do to fix it.

**Comment by chowey** [comment]
@bk2204 That sounds correct to me. I just wanted to reiterate that the workaround from @technoweenie does work:

> As a workaround, I think you should disable smudging completely:
> 
> ```shell
> $ git lfs install --skip-smudge
> $ git lfs env
> ... snip
> git config filter.lfs.process = "git-lfs filter-process --skip"
> git config filter.lfs.smudge = "git-lfs smudge --skip -- %f"
> ```
> 
> After that, you'll have to run `git lfs pull` any time you change branches or fetch updates from your remote.

So it is possible to still work with Git LFS on Windows for large files, but you *must* disable `smudge` or your files will get corrupted.

**Comment by tardyp** [comment]
Hey. This issue should not be closed as the windows client still corrupts files > 4GB.
This is just depending on another bug in git.

**Comment by ttaylorr** [comment]
> This issue should not be closed as the windows client still corrupts files > 4GB.

If I recall correctly, this is not related to a bug in Git, but rather is an inherent limitation of the Windows filesystem.

**Comment by shabbyrobe** [comment]
The issue as described in the Git for Windows bug mentioned earlier in this thread ([git-for-windows/git#1063](https://github.com/git-for-windows/git/issues/1063)) points to a problem with incorrect datatypes in the Git code, not with Windows filesystem limitations:

> The problem is the Git source code, which uses `unsigned long` in places where `size_t` would be correct.

**Comment by bk2204** [comment]
> If I recall correctly, this is not related to a bug in Git, but rather is an inherent limitation of the Windows filesystem.

The limitation on files larger than 4 GiB is for FAT, but not NTFS. NTFS is capable of large files, but you're correct that if you're using a flash drive for your LFS-using Git repository, then you probably have a file system limitation. I think that most Windows users are using NTFS for their systems, though.

I believe in this case the issue is as @shabbyrobe quoted: we use `unsigned long`, which is 32 bit on 64-bit Windows, instead of `size_t`, which is 64 bit. While Unix systems have `unsigned long` equal to the pointer size (they are LP64 systems), Windows always has a 32-bit `long` (the LLP64 model).

There are a handful of patches going into Git 2.21 to address some of these issues, although they may not complete, so it may be useful to keep an eye out for improvements in that regard.

**Comment by aggieNick02** [comment]
I wanted to add that my previous issue with the workaround is no longer present in the latest version of git/git-lfs. So the workaround @technoweenie described has been working great for me. We run a little lfs server (using the lfs-test-server, which is officially non-production, but pretty simple) and have several >4GB files in our repo.

**Comment by aggieNick02** [comment]
FYI, the core issue in git-for-windows (https://github.com/git-for-windows/git/issues/1063) has been closed due to inactivity, but is not fixed. I wanted to try to help others watching this issue and the git-for-windows issue avoid being confused by the closed state change. Utilizing the workaround is probably the best way to deal with this issue for the foreseeable future.

**Comment by velxundussa** [comment]
Hello all,

would putting the workaround ```git lfs pull``` suggested by @technoweenie in a post-checkout hook work correctly or would that skip some cases?

**Comment by bk2204** [comment]
I'm inclined to say adding that to the `post-checkout` hook will likely work, but you may also need to add it to the `post-commit` and `post-merge` hooks to catch all cases (e.g. a merge brings in a change from another branch). Recent versions of Git LFS should honor any changes you make to those hooks unless you run `git lfs update --force`.

**Comment by aggieNick02** [comment]
One 'hole' in the workaround/workflow of configuring `filter.lfs.process`and `filter.lfs.smudge` to be skipped and doing `git lfs pull` remains. As @technoweenie noted, the same limitation does apply to the clean filter, so if you commit and then push a change involving a >4GB file, that file will be malformed after the push. The push will even warn you and direct you to look at smudge's help.

Git will notice that the file is different than it should be if you do a `git status`. A `checkout` and subsequent `git-lfs pull` will fix it. But it would be nice if we could configure `filter.lfs.clean` to be skipped as well; unfortunately it does not have a `--skip` parameter. Would such a a parameter make sense, or are there cases that fall apart? If it did make sense, having it would make for a more robust workflow.

**Comment by bk2204** [comment]
I'm not sure it makes sense, since there's otherwise no way to add a file.  `git lfs checkout` provides a way to check out a file without going through Git, but there's no way to add the LFS contents without doing a `git add`.

**Comment by aggieNick02** [comment]
Ah ok, sorry - I misunderstood the process... I realize skipping `clean` doesn't make sense now. So when I edit an existing >4GB LFS file, commit, and then push, the end result is a corrupt file. Can you help me understand what low-level git-lfs commands are running during that push? `clean` checks file size which seems weird to me since it is turning a file into a pointer, but maybe I'm missing something. I'm trying to understand what command is resulting in the corrupt file so that I can try to avoid that command running, if that makes sense.

**Comment by bk2204** [comment]
There are two possible processes: the `clean` filter (in the config as `filter.lfs.clean`) and the filter process (`filter.lfs.process`).  Those correspond to `git lfs clean` and `git lfs filter-process`.  The latter is used for efficient smudging as well, and is preferred over the actual smudge and clean filters if available.

**Comment by KrisLloyd** [comment]
Been a year since the last update. Has this had any progress?

Running version 2.30.1.windows.1, still receiving errors when using 4GB+ files.

**Comment by bk2204** [comment]
Hey,

As mentioned, this is a limitation in Git for Windows.  Git LFS handles large files just fine without corrupting them on every other platform and the corruption is due to a size problem in Git itself that only manifests on Windows.  If you'd like to see this fixed, please reach out to the Git for Windows team, or better yet, send a patch there or to the Git list.

While I am a contributor to Git, I only do Git development on Linux and as such, I don't believe there are any Git LFS maintainers who have a suitable environment for patching this, so it requires the help of someone who is able and willing to do Git development work on Windows.

**Comment by aggieNick02** [comment]
FYI, the bug in git-for-windows is https://github.com/git-for-windows/git/issues/1063, and has been closed. The git-for-windows maintainer closed the bug (basically as wont-fix) after becoming frustrated with the lack of community contribution towards fixing it and comments asking if progress was being made.

**Comment by bk2204** [comment]
Okay, then it sounds like somebody will need to volunteer to write the patches if they want this fixed.  If someone wants to step up and send patches to the Git list, feel free to CC the address I usually use for development there (check the `.mailmap`) and I'll be happy to chime in with a review.

**Comment by markuskramerIgitt** [comment]
@PhilipOakley wrote [a patch](https://github.com/git-for-windows/git/pull/2179)

**Comment by bk2204** [comment]
That's good to hear.  If there's existing code for this that was tested on Windows at some point, then it is _possible_ that I could rebase it on the latest version and have something that mostly works.  I will, however, note that I have no way of testing things, so even if I did that, things might not work.  I do strongly encourage someone with a suitable Windows system to consider looking at this, since I don't have any time frame for doing it.

**Comment by aggieNick02** [comment]
> That's good to hear.

Kindof... it was a big patch, and, afaict, had several months of involved back and forth review that stopped when the reviewer required the patch be broken down into smaller patches, and the author pushed back on how feasible/meaningful that would or would not be. Then progress stopped around September 2019. So I'm not optimistic about this patch making it in anytime soon.

Just wanted to make that clear so that folks don't get too hopeful that the existence of the patch means the issue is likely to be fixed soon. Obviously it's great that a lot of effort went into fixing this and creating a patch, but the prospects for that work getting incorporated don't seem good.

Unfortunately the workarounds described here may be the only option on windows for a long time.

**Comment by PhilipOakley** [comment]
The main problem with the 4Gb patch(es) is trying to pull it/them apart into sequence of small logical patches that work on both windows and linux both at the same time. 

Most of the (hard) work had already been prepared by others - I managed to work out what to do with zlib itself with a wrapper function.

I never had a method for doing the the patch splitting that would both compile cleanly on linux & windows, and, by the change of a single setting flag, show the next group of incompatibilities that the next patch would solve to maintain that logical sequence. 

I haven't sufficient time to work on Git recently, so it is still on the back burner of considering a  `typedef` for a `memsize` that can do the appropriate tweaks between `long`, `long long`, and `size_t` without the need for dummy casts of the variables, that are added then removed as the feba moves.

**Comment by tardyp** [comment]
Franckly, what is the best value for the community? 
Stick to the split-patches-by-smallest-atomic-commit or fix-the-embarrassing-issue which decrease git adoption rate?

Obviously we don't want to brake any feature, but we mostly care about the final commit, and if we worry about bisect, could we just squash a bit more the patch series for the sake of getting work done?

Thank you very much for your work!

**Comment by PhilipOakley** [comment]
> we mostly care about the final commit, and if we worry about bisect, could we just squash a bit more the patch series

Hi @tardyp 

To clarify, the git maintainer(s) avoid squashing patches, unless really trivial, so they do want to be able to bisect in some detail (it is a Linux core tool after all!). So we will need to ensure they are atomic patches, testable at every step.

(Aside:) There is, in my view, a little bit of difference in the stereotypical programming style between the expectations of the Linux and Windows communities. Whereby the Linux Git community tends to expect a single gifted coder able to do all the work (on their itch) in response to the reviews (though often with suggestions). - A classic 'Open Source' contribution. Meanwhile the Windows programming community (Windows code being often closed source) appears, often, to be more 'team' oriented, so multiple contributors, each with their own particular skill, need to work together to fix perceived issues. Also there appears to be significantly fewer Git-for-Windows contributors, making it harder for a team to coalesce.  

If there is a cohort of folks in the LFS community able to support this particular effort, then - magic!

I've had other engineering work to do, so haven't had much time recently ...

**Comment by tardyp** [comment]
I don't think this is a windows versus linux debate. I am not sure it is a team of average human versus hero developer issue.

There are plenty of classic open source projects that don't use patch series and don't enforce commit level coherency
I maintain one of top ten python projects and we are 1.5 developers working on it with the help of community.
We never use bisect for debugging our issues so we don't bother enforcing commit level coherency.
I am sadden to see you as a developer struggling on maintaining commit level coherency.

What are the odds a git bisect is needed to find a problem in your patch series?
How much time the git bisect user will gain by knowing that its exact problem is located in this 10 line patch while if you squashed, this would be a 1000 lines patch to analyse?
Compare this to the time you spend maintaining your huge patch series as a set of 10 line patches. My guess would be that it is not that useful.

Any way this might not be the right place or community to have this debate :)

**Comment by markuskramerIgitt** [comment]
The above workaround is: 
1) Skip smudge 
2) End user executes `git-lfs pull` 

Do you know of an automated method that replaces step 2?
My intention is that the end user has no need to execute `git-lfs pull`, and can pull from an integrated development environment.


I use this command for cloning:

    GIT_LFS_SKIP_SMUDGE=1 git clone  git@git.fantasy.com:Fantasy/Fantasy-Test.git

The git hook below does not work because the cloning has no hooks yet and git pull has nothing to do (until the repo changed) so the hook is not fired.

```
----snip--------  content of file      .gitconfig     ------snip--------
[core]
	hooksPath = .githook
---------snip-------------------
```


```
----snip--------  content of file      .githooks/post-merge         ------snip--------
#!/bin/sh
echo calling git-lfs pull
git-lfs pull
--snip--------------------------
```
Can post-merge fire also when `git pull` has found no change?

Can a git alias combine git pull and git-lfs pull?

**Comment by PhilipOakley** [comment]
> Any way this might not be the right place or community to have this debate :)

True, but is the 4Gb limit still an issue?

I'm just an Engineer who contributed a part of the effort to extend the zlib wrappers on Git to allow >4Gb when on Windows.

**Comment by markuskramerIgitt** [comment]
@PhilipOakley, unfortunately yes, the 4Gb limit is still an issue,

I first want to find an automated workaround.

**Comment by aggieNick02** [comment]
> @PhilipOakley, unfortunately yes, the 4Gb limit is still an issue,
> 
> I first want to find an automated workaround.

I'm not sure why the post-merge hook isn't running.

Also note that even if it does run, there may be other scenarios (adding a new lfs file, doing a checkout of an older commit hash which has a previous version of a large file) that may still cause problems and not be solved with a post-merge hook. At minimum those scenarios should at least be checked before rolling an automated workaround out to developers with the hopes that the devs won't have to worry about the >4GB issue.

**Comment by markuskramerIgitt** [comment]
@aggieNick02, the post-merge hook seems to be omitted when `git pull` performs  `fast-forward`. If I set `merge.ff false`, the post-merge hook runs, but only at the cost of a merge and a merge message.
In your two scenarios: why would a manually executed `git-lfs pull` create no problems?

**Comment by doctorpangloss** [comment]
I just really want to know what we're supposed to do to use 4GB+ files in Git.

**Comment by doctorpangloss** [comment]
Can `git-lfs` just break up files larger than the 4GB limit into smaller ones, and reassemble them in its smudge or whatever?

**Comment by bk2204** [comment]
> I just really want to know what we're supposed to do to use 4GB+ files in Git.

Well, if you're using Git LFS, you can either use `--skip-smudge` as mentioned above, or you can use WSL to check out the files.  This limitation in Git only applies to Windows versions, not Unix versions.

> Can `git-lfs` just break up files larger than the 4GB limit into smaller ones, and reassemble them in its smudge or whatever?

No, because the problem isn't Git LFS.  The problem is that Git itself will truncate these files, so Git LFS reassembling them will still result in Git truncating them.

**Comment by aggieNick02** [comment]
> Thank you for the follow up. `git lfs install --skip-smudge` doesn't change any behavior for me, in a newly cloned or existing repo. I still receive an error message. On the flip side, my files do not appear to be truncated either way, so this is all very puzzling.
Even when using the workarounds, you will still get warnings from Git LFS about the possible problem. But it typically doesn't actually happen with the workarounds.

**Comment by aggieNick02** [comment]
> @aggieNick02, the post-merge hook seems to be omitted when `git pull` performs `fast-forward`. If I set `merge.ff false`, the post-merge hook runs, but only at the cost of a merge and a merge message.
> In your two scenarios: why would a manually executed `git-lfs pull` create no problems?

@marbx Apologies for not responding months ago - saw and thought about your question with the recent activity.
I don't think a manual `git-lfs pull` would behave any differently than one done automatically via a hook. But I've run into scenarios (like pushing after adding and committing and lfs file, or possibly checking out an older commit) where having the smudge and process filter disabled will put your client in a bad state. There are ways to recover, but the workaround isn't foolproof. Despite that, I've successfully been using Git LFS on windows for at least a couple years with >4GB files, running my own LFS server.

**Comment by marbx** [comment]
Hi @aggieNick02 , thank you for your reply. For me, sadly, foolproof is  mandatory.

I recommended to stop using skip-smudge after it corrupted a ps1 file, maybe because Windows (Server 2012R2) assumed UTF16. Even before, skip-smudge was perceived as process risk.

Until someone (or a group) fixes (or obsoletes) the "Git wrongy assumes that long in C has always 64bit" root cause, 4GB is unavailable for us.

**Comment by PhilipOakley** [comment]
> Until someone (or a group) fixes

If you (or your team, or where you work) are/is in any able to help then that would be useful. 

Even the project scope isn't sufficiently fleshed out to identify a sub-MVP (smallest demonstrable progress - SDP?). I have lot's of personal notes, but it's quite a big job..

**Comment by marbx** [comment]
Hi @PhilipOakley , I guess the team "at hand" is more suitable than that "at work".

I believe I read your rejected PR in which you replaced long with long long. 

I tried to identify a smaller change set by running a debugger with stop points on any long, but I don't found a suitable command to debug. What do you think of that?

Do you have a note about a planned "direct file to file" method to copy files from lfs to git? I only have a vague memory.

**Comment by PhilipOakley** [comment]
Hi @marbx,

My 'at work' aside, was for those readers who might have a work place that would benefit from a >4GB resolution who could maybe get their local management to let them do a few hours on company time, on a win-win basis. Often there is more flexibility than one may imagine (companies cost things funnily;-) 

Which PR were you looking at (so we are talking about the same one). The idea, as per the C89/99 standard, is to use `size_t` as the memsize type (rather than `long long`) so that it is portable between LLP32 and LP64 systems which have a different size for the `int` and `long` types.

I don't have any notes on  '"direct file to file" method to copy files from lfs to git'. I had been just focussed on the internals of Git/Git-for-Windows. 

Having been a Systems Engineer I tend to work from outside in  (big picture), while git patching tends to work from inside to out (fine details picture). So I tend to want to be able to know when the job is complete, rather than small items started.
I hope to show a draft of notes soon once I've added a bit of structure.

One thought of an approachable activity is to look at commands which should NOT involve the 4GB limit and check that it's actually true, or just annotate(split) the [list of commands](https://github.com/git/git/blob/master/command-list.txt) into that same two groups (e.g. shouldn't, maybe, and probably, for 4GB testing - yes that's three, but it should be just two ;-)

**Comment by marbx** [comment]
Yes, I meant your PR with size_t

**Comment by marbx** [comment]
Has anybody else a memory about a direct file to file transfer?

**Comment by KalleOlaviNiemitalo** [comment]
Perhaps it refers to `git lfs checkout`? That would not rely on the smudge filter, I think.

**Comment by PhilipOakley** [comment]
> Yes, I meant your PR with size_t

So that's https://github.com/git-for-windows/git/pull/2179 
I was just making sure we were on the same page.

**Comment by doctorpangloss** [comment]
Would this PR 

> Hi @marbx,
> 
> My 'at work' aside, was for those readers who might have a work place that would benefit from a >4GB resolution who could maybe get their local management to let them do a few hours on company time, on a win-win basis. Often there is more flexibility than one may imagine (companies cost things funnily;-)
> 
> Which PR were you looking at (so we are talking about the same one). The idea, as per the C89/99 standard, is to use `size_t` as the memsize type (rather than `long long`) so that it is portable between LLP32 and LP64 systems which have a different size for the `int` and `long` types.
> 
> I don't have any notes on '"direct file to file" method to copy files from lfs to git'. I had been just focussed on the internals of Git/Git-for-Windows.
> 
> Having been a Systems Engineer I tend to work from outside in (big picture), while git patching tends to work from inside to out (fine details picture). So I tend to want to be able to know when the job is complete, rather than small items started.
> I hope to show a draft of notes soon once I've added a bit of structure.
> 
> One thought of an approachable activity is to look at commands which should NOT involve the 4GB limit and check that it's actually true, or just annotate(split) the [list of commands](https://github.com/git/git/blob/master/command-list.txt) into that same two groups (e.g. shouldn't, maybe, and probably, for 4GB testing - yes that's three, but it should be just two ;-)

I'll happily fix this stuff. Since I live in the here and now I would probably just merge your patch if it passes tests, and otherwise fix the issues the tests show. Based on my cursory reading it sounds like there are some style issues and the patch splitting thing. You don't have to repeat / rehearse for me though.

`git` should work with large files on Windows, there seems to be a belief this isn't a big deal, IMO this is a hair on fire problem with Git-for-Windows. A bajillion things use source control and LFS on large files together like game development and static website content.

**Comment by PhilipOakley** [comment]
> I'll happily fix this stuff. Since I live in the here and now I would probably just merge your patch if it passes tests, and otherwise fix the issues the tests show.

That would be a useful step. Even if you just pull out the zlib changes and (using the git CI) confirm that regular git still works OK, allowing those parts to be [ready for] upstream early.

> You don't have to repeat / rehearse for me though.

That's excellent.


> some style issues and the patch splitting thing.

Yep, that's part of the long-run issue. But if you start somewhere it is still valuable.

**Comment by PhilipOakley** [comment]
@doctorpangloss 

One command that may be worth starting with, as it has a core element to it, is the `git hash-object` command.

Pick the most simple of options to limit the areas of the code base that it could explore (e.g. tweak the `BigFileThreshold` so that you don't have to hit the pack file code). This should at least avoid the combinatorial explosion of interdependencies.

**Comment by mastercoms** [comment]
Is this fixed in Git 2.34?

**Comment by bk2204** [comment]
It is fixed in Git for Windows 2.34, but not in Git 2.34.  The patch was specifically applied to Git for Windows, but has not been released in upstream Git yet.

**Comment by mastercoms** [comment]
Oh I thought that this was a Windows only bug, sorry! Cheers

**Comment by bk2204** [comment]
It is indeed a Windows-only bug, but in the event you're building your own Git on Windows from the Git source, not the Git for Windows source, then there's a difference.  Probably 99% of users on Windows use Git for Windows, so for them, this issue will be fixed in 2.34.

**Comment by bk2204** [comment]
Since this is now fixed upstream, I'm closing this issue.
